snooty/parse_cache.py

```diff
@@ -41,13 +42,34 @@ class CacheData:
     pages: Dict[Tuple[str, str], bytes] = field(default_factory=dict)
     orphan_diagnostics: Dict[str, bytes] = field(default_factory=dict)
 
+    yaml_nodes: Dict[str, Dict[FileId, Tuple[str, bytes]]] = field(
+        default_factory=lambda: collections.defaultdict(dict)
+    )
+    yaml_pages: Dict[str, Sequence[Page]] = field(default_factory=dict)
+
     stats: CacheStats = field(default_factory=CacheStats)
 
     def set_page(self, obj: Page, diagnostics: List[Diagnostic]) -> None:
         self.pages[(obj.ast.fileid.as_posix(), obj.blake2b)] = pickle.dumps(
             (obj, diagnostics), protocol=PROTOCOL
         )
 
+    def ingest_yaml(self, yaml_domain: gizaparser.domain.GizaYamlDomain) -> None:
+        self.yaml_nodes.clear()
+        for category_name, category in yaml_domain.yaml_mapping.items():
+            if category.reified_nodes is None:
+                continue
+
+            for node in category.reified_nodes.values():
+                source_hash = hashlib.blake2b(bytes(node.text, "utf-8")).hexdigest()
+                self.yaml_nodes[category_name][node.path] = (
```

Comment:
I think I may be missing something here, but if we call `clear()` on this dictionary and attempt to assign a nested value, I would expect a `KeyError` to be raised. Apologies if this is not correct, my Python knowledge isn't the best 😅

Reply:
It's a defaultdict! If a key is accessed which doesn't exist, the factory on line 46 (dict()) is called!

Reply:
Ah, that makes sense. Thanks!